### DISCUSSION TITLE
Presenting multiple children from a feature

### DISCUSSION THREAD

**Discussion by ysmaliak** [pointfreeco/swift-composable-architecture#3371] [General]
### Description

There is a bug in the modal presentation behavior when replacing child views. Specifically:

1. When a modal is presented.
2. The modal displays a first child.
3. An attempt is made to replace the first child with a second child.

The replacement operation is not working as expected. The behavior is inconsistent and incorrect, resulting in improper display or failure to update the modal's content.

Note: this is a UIKit project.

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

The first child should be dismissed, and the second child should be presented.

https://github.com/user-attachments/assets/628b4602-b905-448d-a0ae-65c0ccfd0f5a

### Actual behavior

Sometimes (randomly), the parent modal is dismissed when trying to present the second child.

https://github.com/user-attachments/assets/07c0547f-dd8b-40ec-a06e-0121121af3f0

### Steps to reproduce

1. Present Parent as a modal.
2. Present the First Child as a modal from the Parent.
3. Tap a button on the First Child; the Parent should be able to receive this action from the delegate and replace the destination with a Second Child.

Sample project: https://github.com/ysmaliak/tca-navigation-test

### The Composable Architecture version information

1.14.0

### Destination operating system

iOS 17.5

### Xcode version information

15.4

### Swift Compiler version information

```shell
swift-driver version: 1.90.11.1 Apple Swift version 5.10 (swiftlang-5.10.0.13 clang-1500.3.9.4)
Target: arm64-apple-macosx14.0
```

**Comment by mbrandonw** [comment]
Hi @ysmaliak, this is not a bug with the library, but rather just how UIKit (and even SwiftUI) work. You cannot simultaneously dismiss and present controllers, and you can't even simultaneously present two sheets at once. There is even some logs in the console of your app letting you know this wrong:

> Attempt to present <tca_navigation_test.SecondChildViewController: 0x13af15120> on <tca_navigation_test.ParentViewController: 0x13ad0a270> (from <tca_navigation_test.ParentViewController: 0x13ad0a270>) whose view is not in the window hierarchy.

The fix is to space out the `destination` mutations. You can first `nil` out the destination, wait a small amount of time, and then mutate the destination to the next screen.

_However_, I also tested this code in iOS 18.1 and things seem to work as intended. So perhaps UIKit has fixed some bugs. I'm not sure it will always work, but at least this simple use case seems to work as expected.

Since this isn't an issue with the library I am going to convert it to a discussion. Please feel free to continue the conversation over there!

**Reply by ysmaliak** [reply]
Ok, thank you for the explanation and quick response! I will wait some time before presenting the second child.

**Reply by jshier** [reply]
This has come up before, but this definitely seems like something the framework should handle for us. If nothing else, a runtime issue pointing out the issue would be appreciated, as well as a suggested workaround. Ideally we'd be able to chain the dismiss and presentation in some deterministic way (e.g. not a fixed delay). But it'd really be best if the framework just did that for us. (I'd hoped the UIKit integration would do it since I had to do it manually for our existing TCA integration. That it doesn't means we can't move to the native solution.)

**Reply by mbrandonw** [reply]
> This has come up before, but this definitely seems like something the framework should handle for us. If nothing else, a runtime issue pointing out the issue would be appreciated, as well as a suggested workaround.

Do you have any ideas of how to fix this problem at the library level or how to detect it? We would happily take some PR's improving this situation.

**Reply by jshier** [reply]
What we currently do is detect when a presentation in progress while getting a new state, store the computed view controller, and programmatically dismiss the presented controller while tying the presentation of the next controller into the dismiss block of the first. Whether that could work with the native tools I don't know yet.

**Reply by mbrandonw** [reply]
Sounds interesting! Please do let us know if you experiment with how to bake that into the library.

**Reply by jshier** [reply]
@mbrandonw Looking through `swift-navigation`'s implementation, it seems like the core functionality is already there in [`Presentation.swift`](https://github.com/pointfreeco/swift-navigation/blob/d1bdbd8a5d1d1dfd2e4bb1f5e2f6facb631404d4/Sources/UIKitNavigation/Navigation/Presentation.swift#L124). If there's already a `presentedController` that controller is dismissed and then new controller is presented out of the's dismiss' completion. Perhaps TCA just doesn't call into the right overload?

**Reply by mbrandonw** [reply]
@jshier TCA is calling the correct API. You can cmd-click the `present(item:)` in the sample project provided to see ultimately it does get into that exact method (or use a breakpoint to see).

But even with that logic things can still go wrong. The problem is that there is two usages of `present(item:)`:

```swift
present(item: $store.scope(state: \.destination?.firstChild, …
present(item: $store.scope(state: \.destination?.secondChild, …
```

And when setting `destination = .second` both `present(item:)`'s will observe that change. There is a race condition for the ordering. That is why sometimes it works and sometimes it does not.

Really I think only Apple can fix this in UIKit/SwiftUI. They should allow presenting/dismissing as many things as needed in a single tick of the runloop, and then figure out the right thing to do with that information.

**Reply by jshier** [reply]
In general, separate `present`s from the same destination seem like an antipattern for TCA's usage, but I'll have to investigate the example. It still seems like storing some additional state about the presentation would help work around it. Simply saying Apple should fix it isn't really a solution. They haven't fixed this in UIKit in 15 years, they aren't going to.

**Reply by mbrandonw** [reply]
I mentioned this above, but it seems that this is actually fixed in iOS 18. So it does seem like Apple has finally fixed the issue. And if you can sketch out a general solution to this problem we would love to see it.

**Reply by jshier** [reply]
@mbrandonw I'm not sure what you're seeing, but this issue is still trivially reproducible for me in the iOS 18 and 18.1 sims. It's the classic UIKit issue where you try to present something while a dismissal is still underway, and the cause is the lack of synchronization between two `present` methods on the same controller. I'm not sure I'd call it a race condition, as the completion of the dismissal is deterministic, it's just that additional `present` calls need hook into the ongoing dismissal to defer their presentation.

I'm pretty sure this would be solved if TCA offered (and I'm surprised it doesn't) a version of `present` that takes an entire destination at once, similar to the SwiftUI tools.

Alternatively (though really in addition, as I think TCA should offer the functionality), the current `present`s would be updated to handle this case. I'm still thinking about how (probably synchronized through the associated object that was already added), so I don't know if I'll be able to offer a solution directly.

**Reply by stephencelis** [reply]
> I'm pretty sure this would be solved if TCA offered (and I'm surprised it doesn't) a version of `present` that takes an entire destination at once, similar to the SwiftUI tools.

I'm not sure I understand. What tool does SwiftUI offer that we do not?

If you want to unify all your `present`s into a single endpoint that properly handles re-presentation, that is totally doable using an exhaustive enum for the possible presentations:

```swift
present(item: $store.scope(state: \.destination, action: \.destination)) { store in
  switch store.case {
  case .firstChild(let store):
    FirstChildController(store: store)
  case .secondChild(let store):
    SecondChildController(store: store)
  }
}
```

**Reply by mbrandonw** [reply]
@jshier When I ran the demo in iOS 18 it seemed to work just fine, but perhaps I was getting lucky with the race condition I mentioned above. The race condition is that there is no guarantee to the order the `observe`s are called under the hood in `present(item:)`. When it occurs in one order it works, and when it occurs in the other it does not.

> I'm pretty sure this would be solved if TCA offered (and I'm surprised it doesn't) a version of present that takes an entire destination at once, similar to the SwiftUI tools.

This is already possible:

```swift
present(item: $model.destination) { destination in 
  switch destination {
  …
  }
}
```

This just precludes you from having mutually exclusion navigation across sheets/alerts/popovers and drill-downs. You must separate "presentation" destinations from "drill-down" destinations.

And, again, we would love to see what your solution looks like. We would be happy to provide a fix for this issue, but it's not clear what your solution looks like, and sometimes maintaining fixes for UIKit/SwiftUI quirks is more trouble than its worth in the long run.

**Reply by jshier** [reply]
@stephencelis I didn't see a specific overload so I didn't think the functionality existed. I guess it just falls out of the existing API. Unfortunately it also doesn't work, even before the presentation occurs. In the example it instead produces a runtime issue:

> /Users/user/Library/Developer/Xcode/DerivedData/tca-navigation-test-eqeblkobdrjinigzijcdpomodgqj/SourcePackages/checkouts/swift-composable-architecture/Sources/ComposableArchitecture/Reducer/Reducers/Scope.swift:343 A "Scope" at "tca_navigation_test/ParentViewController.swift:14" received a child action when child state was set to a different case. …
>
>  Action:
>    ParentFeature.Destination.Action.firstChild(.buttonTapped)
>  State:
>    ParentFeature.Destination.State.secondChild
>
>This is generally considered an application logic error, and can happen for a few reasons:
>
> • A parent reducer set "ParentFeature.Destination.State" to a different case before the scoped reducer ran. Child reducers must run before any parent reducer sets child state to a different case. This ensures that child reducers can handle their actions while their state is still available. Consider using "Reducer.ifCaseLet" to embed this child reducer in the parent reducer that change its state to ensure the child reducer runs first.
>
> • An in-flight effect emitted this action when child state was unavailable. While it may be perfectly reasonable to ignore this action, consider canceling the associated effect before child state changes to another case, especially if it is a long-living effect.
>
> • This action was sent to the store while state was another case. Make sure that actions for this reducer can only be sent from a view store when state is set to the appropriate case. In SwiftUI applications, use "SwitchStore".

@mbrandonw That's interesting, I didn't realize the order of observations was non deterministic. Explains why changing the order of the `present`s didn't improve anything.

I'm still pretty sure that if `present`ing the `destination` directly worked, the underlying UIKit integration would properly dismiss and present and this issue would be solved, users would just have to make sure to use that form.

**Reply by jshier** [reply]
Actually, that runtime issue only comes up when you tap the button a second time, as it's sending the action after the `destination` has changed. The first tap changes the state but doesn't do anything. I'll investigate that part.

**Reply by jshier** [reply]
When scoped to the `destination`, `present` isn't called when the state transitions directly to the second state. Investigating, it looks like the root `observe` in the parent exits early because the `id` of the `item` hasn't changed, despite the underlying state changing. I *think* this is due the underlying `item` being the same `Store` instance, which generates the same `id` every time since the reference, of course, hasn't changed. Whether this is desirable or correct I can't say yet, but I think it means any reference type used in these tools will have the same problem.

**Reply by FredericRuaudel** [reply]
Hi ! 

If it can help, I've also made a project that showcase this kind of navigation in both UIKit and SwiftUI with both usecases: setting the new destination directly and setting destination to nil, waiting then setting the second destination. 

In SwiftUI, only the first usecase is working while in UIKit both fail

https://github.com/FredericRuaudel/DoubleModaleShowCase

I also try to run the project with the draft PR that seems related https://github.com/pointfreeco/swift-composable-architecture/pull/3464 but with no luck, it behaves exactly the same.

**Reply by FredericRuaudel** [reply]
As a follow-up, I finally fix my issue by simply increasing the waiting time between the set of the first popover to nil and the set of it to the second destination from 300ms to 1sec. 500ms was not enough but I didn't try to fine-tune more than that so maybe we can find a lesser value between 500ms and 1sec. 
The fix is available in the last commit of my project linked above.

To sum it up, it goes like this:

```swift
           switch action {
           // …
           case let .destination(.presented(.shareOptions(.delegate(delegateAction)))):
                switch delegateAction {
                case let .dismissThenShare(value):
                    state.destination = nil
                    return .run { [value] send in
                        try await Task.sleep(nanoseconds: 1000 * NSEC_PER_MSEC) // Here we need to wait more than 500ms at least
                        await send(.showShareSheet(value))
                    }
                }

            case .showShareSheet(let value):
                state.destination = .shareSheet(.init(value: value))
                return .none
            }
```

**Comment by FredericRuaudel** [comment]
It seems that the issue has worsened in iOS 26. I've opened issue https://github.com/pointfreeco/swift-composable-architecture/issues/3836 with sample code to showcase the problem
